controllers/btpoperator_controller.go

```diff
@@ -1131,3 +1201,480 @@ func (r *BtpOperatorReconciler) watchConfigPredicates() predicate.Funcs {
 		UpdateFunc: func(e event.UpdateEvent) bool { return nameMatches(e.ObjectNew) },
 	}
 }
+
+func (r *BtpOperatorReconciler) watchWebhooksPredicates() predicate.Funcs {
+	predicateForCertificates := func(obj client.Object) bool {
+		switch v := obj.(type) {
+		case *admissionregistrationv1.ValidatingWebhookConfiguration:
+			{
+				return v.Name == validatingWebhookName
+			}
+		case *admissionregistrationv1.MutatingWebhookConfiguration:
+			{
+				return v.Name == mutatingWebhookName
+			}
+		default:
+			return false
+		}
+	}
+
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return predicateForCertificates(e.Object)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return predicateForCertificates(e.Object)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return predicateForCertificates(e.ObjectOld)
+		},
+	}
+}
+
+func (r *BtpOperatorReconciler) reconcileCertificates(ctx context.Context, resourcesToApply *[]*unstructured.Unstructured) error {
+	logger := log.FromContext(ctx)
+	logger.Info("checking certifications started")
+
+	caSecretExists, err := r.checkIfSecretExists(ctx, CaSecret)
+	if err != nil {
+		return err
+	}
+	if !caSecretExists {
+		logger.Info("ca secret with cert dosent exists.")
+		if err := r.doFullyCertificatesRegenerate(ctx, resourcesToApply); err != nil {
+			return err
+		}
+		return nil
+	}
+	logger.Info("ca secret exists")
+
+	webhookSecretExists, err := r.checkIfSecretExists(ctx, WebhookSecret)
+	if err != nil {
+		return err
+	}
+	if !webhookSecretExists {
+		logger.Info("web secret secret with cert dosent exists.")
+		if err := r.doPartialCertificatesRegenerate(ctx, resourcesToApply); err != nil {
+			return err
+		}
+		if err := r.reconcileWebhooks(ctx, resourcesToApply, nil); err != nil {
+			return err
+		}
+		return nil
+	}
+	logger.Info("webhook secret exists")
+
+	if err := r.reconcileWebhooks(ctx, resourcesToApply, nil); err != nil {
+		return err
+	}
+	logger.Info("webhooks ca bundles are ok")
+
+	signOk, err := r.isWebhookSecretCertSignedByCaSecretCert(ctx)
+	if err != nil {
+		logger.Error(err, "while checking if webhook is signed by correct ca")
+		if err := r.doFullyCertificatesRegenerate(ctx, resourcesToApply); err != nil {
+			return err
+		}
+		return nil
+	}
+	if !signOk {
+		logger.Error(nil, "webhook cert is not signed by correct ca")
+		if err := r.doFullyCertificatesRegenerate(ctx, resourcesToApply); err != nil {
+			return err
+		}
+		return nil
+	}
+	logger.Info("webhook certificate is signed by correct root")
+
+	caCertificateValid, err := r.doesCertificateExpireSoon(ctx, CaSecret)
```

Comment:
The naming is misleading - function checks if certificate expires soon not if it is valid.
Not expired certificate (which is supposed to expire soon) is valid.

Reply:
And if expires soon we decided not to use it but to regenerate.